firedrake/mg/ufl_utils.py

```diff
@@ -126,33 +126,14 @@ def coarsen_bc(bc, self, coefficient_mapping=None):
     return type(bc)(V, val, subdomain)
 
 
-@coarsen.register(firedrake.functionspaceimpl.FiredrakeDualSpace)
-@coarsen.register(firedrake.functionspaceimpl.FunctionSpace)
-@coarsen.register(firedrake.functionspaceimpl.WithGeometry)
+@coarsen.register(firedrake.functionspaceimpl.WithGeometryBase)
 def coarsen_function_space(V, self, coefficient_mapping=None):
     if hasattr(V, "_coarse"):
         return V._coarse
-    fine = V
-    indices = []
-    while True:
-        if V.index is not None:
-            indices.append(V.index)
-        if V.component is not None:
-            indices.append(V.component)
-        if V.parent is not None:
-            V = V.parent
-        else:
-            break
 
+    fine = V
     mesh = self(V.mesh(), self)
```

Comment:
Redefining `mesh` and `V` is confusing. New names for the coarse objects would make this code easier to parse.
```suggestion
    V_fine = V
    mesh_coarse = self(V_fine.mesh(), self)
```